packages/htmlviewer/src/index.ts

```diff
@@ -0,0 +1,146 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) Jupyter Development Team.
+| Distributed under the terms of the Modified BSD License.
+|----------------------------------------------------------------------------*/
+
+import { IFrame } from '@jupyterlab/apputils';
+
+import { ActivityMonitor } from '@jupyterlab/coreutils';
+
+import {
+  ABCWidgetFactory,
+  DocumentRegistry,
+  DocumentWidget,
+  IDocumentWidget
+} from '@jupyterlab/docregistry';
+
+import '../style/index.css';
+
+/**
+ * The timeout to wait for change activity to have ceased before rendering.
+ */
+const RENDER_TIMEOUT = 1000;
+
+/**
+ * The CSS class to add to the HTMLViewer Widget.
+ */
+const CSS_CLASS = 'jp-HTMLViewer';
+
+/**
+ * A viewer widget for HTML documents.
+ *
+ * #### Notes
+ * The iframed HTML document can pose a potential security risk,
+ * since it can execute Javascript, and make same-origin requests
+ * to the server, thereby executing arbitrary Javascript.
+ *
+ * Here, we sandbox the iframe so that it can't execute Javsacript
+ * or launch any popups. We allow one exception: 'allow-same-origin'
```

Comment:
allow-same-origin allows document traversal between pages with the same origins- it does not (AFAIK) impact resource loading within the iframe. If it is indeed not strictly needed then it's probably better omitted as it becomes a more significant issue if combined with allow-scripts, or maybe clarifying.

Reply:
Thanks for the information @blois! i'd certainly like to make this as strict as possible, so if resource loading is not affected here, I'd just as soon remove the exception.

I'm currently not allowing scripts, but am trying to think through what the circumstances might need to be to allow them. I am currently thinking something around explicit user trusting of the file may be okay. It's hard to be *completely* safe, since the whole point of the application is running code on the server. Do you have any insights there?

Reply:
Since you're just assigning it via a Blob then using a srcdoc iframe along with the sandbox attribute is probably the best, as it's the easiest way to give your iframe a separate origin.

Reply:
In my testing there were a number of fetches of local resources that failed when I didn't include `'allow-same-orign'`, so I'm concerned that's *too* strict of a sandboxing.